sources/packages/api/src/services/sin-validation/sin-validation.service.ts

```diff
@@ -1,80 +1,195 @@
+        "sinValidation.dobSent",
+        "sinValidation.genderSent",
+        "sinValidation.temporarySIN",
+        "sinValidation.sinExpireDate",
+        "user.id",
+      ])
+      .innerJoin("sinValidation.user", "user")
+      .where("sinValidation.id = :sinValidationId", {
+        sinValidationId: validationResponse.referenceIndex,
+      })
+      .getOne();
+
+    if (!existingValidation) {
+      throw new Error(
+        `Not able to find the SIN validation id ${validationResponse.referenceIndex} to be updated with the ESDC response.`,
+      );
+    }
+
+    const sinValidationNeverUpdated = !existingValidation.dateReceived;
+    const sinStatusChanged =
+      existingValidation.sinStatus !== validationResponse.sinCheckStatus;
+
+    // Values to be updated from the ESDC response.
+    existingValidation.dateReceived = processDate;
+    existingValidation.fileReceived = receivedFileName;
+    existingValidation.isValidSIN = isValidSIN;
+    existingValidation.sinStatus = validationResponse.sinCheckStatus;
+    existingValidation.validSINCheck = validationResponse.sinOkayFlag;
+    existingValidation.validBirthdateCheck =
+      validationResponse.birthDateOkayFlag;
+    existingValidation.validFirstNameCheck =
+      validationResponse.firstNameOkayFlag;
+    existingValidation.validLastNameCheck = validationResponse.lastNameOkayFlag;
+    existingValidation.validGenderCheck = validationResponse.genderOkayFlag;
+    existingValidation.modifier = { id: auditUserId } as User;
+
+    if (sinValidationNeverUpdated) {
+      // The SIN validation record was never updated before,
+      // update it with the information received from ESDC.
+      // This will be most common scenario.
+      operationDescription = "SIN validation record updated.";
+      const updatedRecord = await this.repo.save(existingValidation);
+      return {
+        operationDescription,
+        record: updatedRecord,
+      };
+    }
+
+    if (sinStatusChanged) {
+      // The SIN validation is already present and the received status is different from
+      // the existing one. It can happen when the status received is 2 (under review) and later
+      // a new status will be received once the review is done, what means that one request can
+      // result in more than one response. In this case, if the received record is the most updated
+      // for the student, a "cloned" record will be inserted.
+      const mostUpdatedRecordDate = await this.repo
+        .createQueryBuilder("sinValidation")
+        .select("MAX(sinValidation.dateReceived)", "maxDate")
+        .innerJoin("sinValidation.user", "user")
+        .where("user.id = :userId", {
+          userId: existingValidation.user.id,
+        })
+        .getRawOne();
+      if (processDate > mostUpdatedRecordDate.maxDate) {
+        // This will force the record to be inserted with all the values selected
+        // during the initial SQL query.
+        delete existingValidation.id;
```

Comment:
should we also need to delete the previous or old audit fields and add new audit fields for the new cloned row?

Reply:
The only columns retrieved in the SQL select are the ones meant to be cloned. I added comments about it on the query and in line 163 also "to be inserted with all the values selected during the initial SQL query."